docs/tpm-spi-dev.md

```diff
@@ -0,0 +1,398 @@
+> filled buffer, and using HAL_SPI_TransmitReceive instead of HAL_SPI_Receive.
+> Zephyr (see the section below) already does that which partially solves the
+> problem. However, if TPM does not enqueue transfer right in time, this causes
+> noise on SPI.
+
+And now, execute command:
+
+```shell
+$ spitest -v -D /dev/spidev1.0 -s 100000 -p '\x01\x01\x00\x00\x01'
+```
+
+You should see something like this on SWV console
+
+```shell
+TPM command: 80 01 00 00 00 0c 00 00 01 44 00 00
+2000.01.01-00:20:18.201GMT: Executing command TPM_CC_Startup()
+2000.01.01-00:20:18.525GMT: NVFile written (16kb, 1970.01.01-00:00:00GMT created, 0 writes, NEVER last)
+2000.01.01-00:20:18.525GMT: Completion time 0'0" with ReturnCode {TPM_RC_SUCCESS}
+Response size: 10
+```
+
+To read response, type:
+
+```shell
+$ spitest -v -D /dev/spidev1.0 -s 100000 -p '\x04\x04\x00\x00\x00\x00\x00\x00'
+$ spitest -v -D /dev/spidev1.0 -s 100000 -p '\x00\x00\x00\x00'
+spi mode: 0x0
+bits per word: 8
+max speed: 100000 Hz (100 kHz)
+TX | 00 00 00 00 __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __  |....|
+RX | 0A 00 00 00 __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __  |....|
+```
+
+Response size is 10, so we need to send 4 byte header and then prepare 10 byte
+transfer
+
+```shell
+$ spitest -v -D /dev/spidev1.0 -s 100000 -p '\x04\x02\x00\x00\x00\x00\x00\x00'
+$ spitest -v -D /dev/spidev1.0 -s 100000 -p '\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
+spi mode: 0x0
+bits per word: 8
+max speed: 100000 Hz (100 kHz)
+TX | 00 00 00 00 00 00 00 00 00 00 __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __  |..........|
+RX | 80 01 00 00 00 0A 00 00 00 00 __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __ __  |..........|
+```
+
+Examples above show successful communication attempt, however communication is
+unstable and it tends to fail frequently. Failure can be simply reproduced by
+trying commands multiple times.
+
+Also, please note that SPI is running at 100 kHz, where TPM specification
+requires TPM to operate at range 1 MHz - 24 MHz. Currently, upon receiving
+command TPM is waiting for host to receive response and will not accept further
+commands until response is received. Before sending command and receiving
+response there is a delay caused by executing `spitest` separately. `spitest`
+calls may be chained together, but this may result in TPM not responding right
+in time.
+
+STM32 HAL does not provide API to sample SPI chip select, but TPM specification
+requires some actions to be taken based on the CS value.
+
+> From [TCG PC Client Platform TPM Profile Specification](https://trustedcomputinggroup.org/wp-content/uploads/PC-Client-Specific-Platform-TPM-Profile-for-TPM-2p0-v1p05p_r14_pub.pdf)
+> section 6.5.1.1:
+>
+> 1. For Read Cycles, the TPM SHALL abort a cycle by driving 1 on MISO and
+>    continue to hold MISO at 1 until its CS# signal is deasserted.
```

Comment:
Theoretically yes, but we still need to monitor CS# to know when to stop sending FFs. Otherwise after platform re-asserted CS# we would still continue sending FFs.